src/libcore/slice.rs

```diff
@@ -684,13 +666,11 @@ macro_rules! iterator {
             fn next(&mut self) -> Option<$elem> {
                 // could be implemented with slices, but this avoids bounds checks
                 unsafe {
-                    ::intrinsics::assume(!self.ptr.is_null());
-                    ::intrinsics::assume(!self.end.is_null());
```

Comment:
@bluss Ok, I wasn't sure what actually happened if it overflowed. I assumed that the resulting value couldn't actually be dereferenced, but I thought it was plausible that it could still be used as a pointer comparison (since one-past-the-end is a valid location, and having an object occupy the last byte in the address space seems like a plausible thing to support). But if the documentation explicitly says it's a poison value then I guess it is.

That said, this only applies to slices of non-zero-sized elements. `offset()` on a slice of zero-sized elements results in the same pointer (because anything multiplied by `0` is still `0`). The iterators use that macro specifically to get around this case, but e.g. indexing into a slice of zero-sized elements uses `offset()` and that's fine (it just results in the same pointer that it was called on).

In light of this, I can re-add the `end` assumption to `next()`.